compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt

```diff
@@ -0,0 +1,405 @@
+     * that different stages of animations spawned by this effect for both dimensions
+     * end at the same time, which is not the case:
+     * Spring->Fling->Spring, Fling->Spring, Spring->Fling effects can have different timing per dimension
+     * (see Notes of https://github.com/JetBrains/compose-multiplatform-core/pull/609),
+     * which is not possible to express without changing API. Hence this effect will be fixed to latest
+     * received delta.
+     */
+    private var direction: CupertinoOverscrollDirection = CupertinoOverscrollDirection.UNKNOWN
+
+    private val reverseHorizontal =
+        when (layoutDirection) {
+            LayoutDirection.Ltr -> false
+            LayoutDirection.Rtl -> true
+        }
+
+
+    /*
+     * Size of container is taking into consideration when computing rubber banding
+     */
+    private var scrollSize: Size = Size.Zero
+
+    /*
+     * Current offset in overscroll area
+     * Negative for bottom-right
+     * Positive for top-left
+     * Zero if within the scrollable range
+     * It will be mapped to the actual visible offset using the rubber banding rule inside
+     * [Modifier.offset] within [effectModifier]
+     */
+    private var overscrollOffset: Offset by mutableStateOf(Offset.Zero)
+
+    private var lastFlingUncosumedDelta: Offset = Offset.Zero
+    private val visibleOverscrollOffset: IntOffset
+        get() =
+            overscrollOffset.reverseHorizontalIfNeeded().rubberBanded().round()
+
+    override val isInProgress: Boolean
+        get() =
+        // If visible overscroll offset has at least one pixel
+            // this effect is considered to be in progress
+            visibleOverscrollOffset.toOffset().getDistance() > 0.5f
+
+    override val effectModifier = Modifier
+        .onPlaced {
+            scrollSize = it.size.toSize()
+        }
+        .offset {
+            visibleOverscrollOffset
+        }
+
+//    internal suspend fun playInitialSpringIfNeeded(velocity: Float): Float {
+//        val scrollValueConverter = scrollValueConverter ?: return velocity
+//
+//        val currentOverscroll = scrollValueConverter.convertOffsetToFloat(overscrollOffset)
+//
+//        return if ((velocity > 0f && currentOverscroll > 0f) || (velocity < 0f && currentOverscroll < 0f)) {
+//            playSpringAnimation(0f, velocity, true)
+//        } else {
+//            velocity
+//        }
+//    }
+
+    private fun NestedScrollSource.toCupertinoScrollSource(): CupertinoScrollSource? =
+        when (this) {
+            NestedScrollSource.Drag -> CupertinoScrollSource.DRAG
+            NestedScrollSource.Fling -> CupertinoScrollSource.FLING
+            else -> null
+        }
+
+    /*
+     * Takes input scroll delta, current overscroll value, and scroll source. Returns pair of
+     * 1. Available delta to perform actual content scroll.
+     * 2. New overscroll value.
```

Comment:
Create 1-liner data class for such things

Reply:
Done